File: leetcode/algorithms/c++/Longest Palindromic Substring/Longest Palindromic Substring.cpp
class Solution
{
private:
    int n;
    string str;
    vector<vector<int>> dp;
    vector<vector<bool>> vis;

    int dfs(int x, int y)
    {
        if (x < 0 || y < 0 || x >= n || y >= n || x > y)
            return 0;

        if (vis[x][y])
            return dp[x][y];
        vis[x][y] = true;

        if (str[x] == str[y])
        {

            int ans = dfs(x + 1, y - 1);
            return dp[x][y] = ans > 0 ? ans + 2 : 0;
        }
        return 0;
    }

public:
    string longestPalindrome(string s)
    {
        n = s.size();
        if (n == 0)
            return "";

        str = s;
        dp.resize(n);
        vis.resize(n);
        for (int i = 0; i < n; i++)
        {
            dp[i].resize(n);
            vis[i].resize(n);
            for (int j = 0; j < n; j++)
            {
                dp[i][j] = 0;
                vis[i][j] = false;
            }
        }

        int len = -1;
        int x = -1;
        int y = -1;

        for (int i = 0; i < n; i++)
        {
            vis[i][i] = true;
            dp[i][i] = true;
            if (i + 1 < n && s[i] == s[i + 1])
            {
                vis[i][i + 1] = true;
                dp[i][i + 1] = 2;
            }
        }

        for (int i = 0; i < n; i++)
        {
            for (int j = i; j < n; j++)
            {
                dp[i][j] = dfs(i, j);
                if (dp[i][j] > len)
                {
                    x = i;
                    y = j;
                    len = dp[i][j];
                }
            }
        }

        string res = "";
        for (int i = x; i <= y; i++)
            res += str[i];

        return res;
    }
};

File: leetcode/algorithms/python/Longest Palindromic Substring/Longest Palindromic Substring.py
class Solution:
    
    n = 0
    vis = []
    dp = []
    word = ""

    def dfs(self, x: int, y: int) -> int:
        if (x < 0 or y < 0 or y >= self.n or x >= self.n or x > y):
            return 0
        if (self.vis[x][y] == True):
            return self.dp[x][y]
        self.vis[x][y] = True
        if (self.word[x] == self.word[y]):
            ans = self.dfs(x + 1, y - 1)
            if (ans > 0):
                self.dp[x][y] = ans + 2
            else:
                self.dp[x][y] = 0
        return 0


    def longestPalindrome(self, s: str) -> str:
        self.n = len(s)
        if (self.n == 0):
            return ""
        self.word = s
        for i in range(self.n):
            self.dp.append([])
            self.vis.append([])
            for j in range(self.n):
                self.dp[i].append(0)
                self.vis[i].append(False)
        length = -1
        x = -1
        y = -1
        for i in range(self.n):
            self.vis[i][i] = True
            self.dp[i][i] = 1
            if (i + 1 < self.n and s[i] == s[i + 1]):
                self.vis[i][i+1] = True
                self.dp[i][i + 1] = 2
        for i in range(self.n):
            for j in range(self.n):
                self.dp[i][j] = self.dfs(i, j)
                if (self.dp[i][j] > length):
                    x = i
                    y = j
                    length = self.dp[i][j]
        return self.word[x:y]